### ISSUE TITLE
ELECTRA: Pre-training Text Encoders as Discriminators Rather Than Generators

### ISSUE THREAD

**Issue by aisolab** [aisolab/paper-log#1]
## 어떤 내용의 논문인가요? 👋
- 본 논문의 기존의 pretrained language model 방법론들이 pretraining 시 많은 computing resource를 요구하는 점을 개선하기위해, sample-efficient한 pretraining 방법론인 replaced token detection을 제안하는 논문입니다.
- masked language model (BERT 계열), permutation language model (XLNET) 과는 다르게, transformer encoder로 각각 generator와 discriminator를 구성합니다.
- generator는 기존의 BERT 계열과 비슷하게 masked input에 대하여 masked language model의 loss로 backprop을 한 뒤 (15% masking), masked token에 대하여 token을 sample해서 corrupted input을 구성합니다. (eg. 논문에서는 여러가지 training 방법을 제안하고 있지만, 가장 default한 것을 적었습니다.)
- discriminator는 corrupted input의 모든 token에 대하여 각각의 token들이 sample된 것 인지, 원래 input에 있던 것인지를 구별하는 loss (real, fake prediction)를 통해 학습합니다
  - masked language model이 input의 subset tokens of input들에 대해서 loss를 계산하는 반면, 이 방법은 all tokens of input에 대해서 loss를 계산하므로 sample-efficient함
  - 또한 discriminator 입장에서 pretraining과 finetuning시에 discrepancy가 없음
  - 해당 loss는 generator에 backprop되지 않습니다.  이 때 sampled token이 원래 input의 token과 똑같다면, real로 간주합니다.
- pretraining이 완료되면 generator는 버리고, discriminator만 finetuning해서 사용합니다.
## Abstract (요약) 🕵🏻‍♂️
Masked language modeling (MLM) pre-training methods such as BERT corrupt the input by replacing some tokens with [MASK] and then train a model to reconstruct the original tokens. While they produce good results when transferred to downstream NLP tasks, they generally require large amounts of compute to be effective. As an alternative, ***we propose a more sample-efficient pre-training task called replaced token detection. Instead of masking the input, our approach corrupts it by replacing some tokens with plausible alternatives sampled from a small generator network.*** Then, instead of training a model that predicts the original identities of the corrupted tokens, ***we train a discriminative model that predicts whether each token in the corrupted input was replaced by a generator sample or not.*** Thorough experiments demonstrate ***this new pre-training task is more efficient than MLM because the task is defined over all input tokens rather than just the small subset that was masked out.*** As a result, ***the contextual representations learned by our approach substantially outperform the ones learned by BERT given the same model size, data, and compute.*** The gains are particularly strong for small models; for example, we train a model on one GPU for 4 days that outperforms GPT (trained using 30x more compute) on the GLUE natural language understanding benchmark. ***Our approach also works well at scale, where it performs comparably to RoBERTa and XLNet while using less than 1/4 of their compute and outperforms them when using the same amount of compute.***
## 이 논문을 읽어서 무엇을 배울 수 있는지 알려주세요! 🤔
- Contrastive Learning에 대한 일반론을 배울 수 있습니다.
## 같이 읽어보면 좋을 만한 글이나 이슈가 있을까요?
- [BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/abs/1810.04805)
- [XLNet: Generalized Autoregressive Pretraining for Language Understanding](https://arxiv.org/abs/1906.08237)
## 레퍼런스의 URL을 알려주세요! 🔗
https://openreview.net/forum?id=r1xMH1BtvB&utm_campaign=piqcy&utm_medium=email&utm_source=Revue%20newsletter

**Comment by aisolab** [comment]
# Motivation
- sample-efficient한 pretrained language model 학습 방법론을 제안합니다.
- 기존의 pretrained language model (eg. BERT, XLNet)이 input의 subset tokens of input들로만 학습하고 있음을 개선하려고 합니다.
# Method
![image](https://user-images.githubusercontent.com/19755607/75618052-adb3d300-5bab-11ea-8655-0d723393a034.png)
- masked language model의 loss (masked token이 원래 무슨 token인지 파악)로 back propagation이 되는 small generator (transformer encoder)가 masked token에 대해서 token을 sampling합니다.
- discriminator (transformer encoder)가 모든 token에 대해서 contrastive learning의 loss (token이 원래 input에 있던 token과 같은 지 또는 다른 지 파악)를 계산하고 back propagation합니다
- 다양한 training 방법을 제안하나 본 논문에서 가장 좋은 joint-training 방식 (generator와 discriminator 동시 학습)을 활용하여 pretraining을 합니다.
- pretraining이 완료되면 discriminator만 가지고 task에 맞추어 적절히 finetuning하여 사용합니다.
# Experiment
![image](https://user-images.githubusercontent.com/19755607/75618113-a4773600-5bac-11ea-8b19-0884bb1a2f42.png)
![image](https://user-images.githubusercontent.com/19755607/75618135-f8821a80-5bac-11ea-915b-30299373b666.png)
- table 2 (glue dev)를 보면 flops 수 기준으로 RoBERTa-500K, XLNET의 1/4 정도만 학습해도 comparable한 performance를 가집니다.
- table 3 (glue test)를 보면 동일 flops에서 sota 급의 성능을 가집니다.
- table 4 (squad)에 경우에도 적은 flops에서 ALBERT를 제치는 sota를 달성합니다.
# Comment
- subset tokens of input에 대해서 pretraining을 진행하는 것을 all tokens of input으로 대체하는 것이 성능 향상에 크게 기여합니다.
- pretraining과 finetuning의 mismatch ([MASK} token 사용)을 개선하는 것도 미미하지만 성능 향상에 기여합니다.
